niceman/retrace/packagemanagers.py

```diff
@@ -253,14 +253,21 @@ def _create_package(self, pkgname):
         for v in pkg_info.versions:
             v_info = {"version": v.version}
             origins = []
-            for o in v.origins:
-                origin = {"component": o.component,
-                          "archive": o.archive,
-                          "origin": o.origin,
-                          "label": o.label,
-                          "site": o.site}
+            for (pf, _) in v._cand.file_list:
+                # Pull origin information from package file
+                origin = {"component": pf.component,
+                          "archive": pf.archive,
+                          "origin": pf.origin,
+                          "label": pf.label,
+                          "site": pf.site}
+                # Get the archive uri
+                indexfile = v.package._pcache._list.find_index(pf)
+                if indexfile:
+                    archive_uri = indexfile.archive_uri("")
+                    origin["archive_uri"] = archive_uri
                 # Get the release date
-                rdate = self._find_release_date(o.site, o.archive)
+                rdate = self._find_release_date(
+                    self._find_release_file(pf.filename))
                 if rdate:
                     origin["date"] = rdate
```

Comment:
You are correct, I have misread... In my PR I had the same mental picture as you but didn't want to deal with every setting this way so came up with a helper to prune all empty ones at the end.. have a look, we should merge the efforts ;-)